### ISSUE TITLE
Warning: findDOMNode is deprecated in StrictMode

### ISSUE THREAD

**Issue by wildseeder** [arco-design/arco-design#81]
- [x] I'm sure this does not appear in [the issue list of the repository](https://github.com/arco-design/arco-design/issues)

## Basic Info
- **依赖包名 及 版本:** @arco-design/web-react: 2.23.5
- **系统版本:** all
- **浏览器版本:** all
- **复现地址:** https://codepen.io/wildseeder/pen/ZEJJNvd

## 预期结果
No warning messages in console.

## 复现步骤
Use menu component when React strict mode is enable, there are "findDOMNode is deprecated in StrictMode" warning messages in console.

**Comment by PengJiyuan** [comment]
Yes, this is a known serious problem. At present, in some components, findDOMNode needs to be used to find component dom nodes. There is no alternative method for the time being. React officials do not provide a solution to completely replace findDOMNode. We need to further study

**Comment by claneo** [comment]
@PengJiyuan 在源码里搜了一下，findDOMNode 出现的次数也不算太多，看起来都能处理一下

1. 有个叫 getNodeHeight 的函数，看了下传参已经是一个 HTMLElement 了，没必要用 findDOMNode 了
2. ResizeObserverComponent 是个内部的组件，应该可以直接换成读 children 的 ref
3. 有些组件里用到了 findDOMNode(getPopupContainer())，这里可以判断下，如果 getPopupContainer 返回的是个 ReactElement 再使用 findDOMNode，并在文档里提示
4. 最后一个就是 Trigger 了，首先里面有一半是 findDOMNode(this.triggerRef)，可以直接去掉。然后麻烦的就是获取 children 的 dom 了

目前 mui 的做法是强制 children 必须有 ref

这里感觉可以改成先读取 children 的 ref，如果拿到的不是 HTMLElement 的话，再使用 findDOMNode

对于像 Input 这样 ref 返回的不是个 dom 的情况，要么是让用户自己用一个新的组件包一层，要么是 Trigger 支持下自定义从 ref 获取 dom

比如这样

方式1，缩进会少一点

```jsx
<Trigger getDomFromRef={ref => ref.dom}>
    <Input />
</Trigger>
```

方式2，缩进有点多，但是更灵活，虽然我也不知道什么情况下会这么写，明明把 Trigger 直接写里面就好

```jsx
<Trigger>
    {(getDomFromRef) => (
        <>
            <Input ref={(ref) => getDomFromRef(ref.dom)} />
        </>
    )}
</Trigger>;
```

另外可能还需要在文档里注明，都有哪些组件 ref 返回的不是个 dom，比如 Input

**Comment by liran319** [comment]
I met the same warning about this. Is there any update?

**Comment by chenc041** [comment]
> I met the same warning about this. Is there any update?

no。。。😂

**Comment by Galileo01** [comment]
Is there any progress on this issue?

**Comment by PengJiyuan** [comment]
> @PengJiyuan 在源码里搜了一下，findDOMNode 出现的次数也不算太多，看起来都能处理一下
> 
> 1. 有个叫 getNodeHeight 的函数，看了下传参已经是一个 HTMLElement 了，没必要用 findDOMNode 了
> 2. ResizeObserverComponent 是个内部的组件，应该可以直接换成读 children 的 ref
> 3. 有些组件里用到了 findDOMNode(getPopupContainer())，这里可以判断下，如果 getPopupContainer 返回的是个 ReactElement 再使用 findDOMNode，并在文档里提示
> 4. 最后一个就是 Trigger 了，首先里面有一半是 findDOMNode(this.triggerRef)，可以直接去掉。然后麻烦的就是获取 children 的 dom 了
> 
> 目前 mui 的做法是强制 children 必须有 ref
> 
> 这里感觉可以改成先读取 children 的 ref，如果拿到的不是 HTMLElement 的话，再使用 findDOMNode
> 
> 对于像 Input 这样 ref 返回的不是个 dom 的情况，要么是让用户自己用一个新的组件包一层，要么是 Trigger 支持下自定义从 ref 获取 dom
> 
> 比如这样
> 
> 方式1，缩进会少一点
> 
> ```js
> <Trigger getDomFromRef={ref => ref.dom}>
>     <Input />
> </Trigger>
> ```
> 
> 方式2，缩进有点多，但是更灵活，虽然我也不知道什么情况下会这么写，明明把 Trigger 直接写里面就好
> 
> ```js
> <Trigger>
>     {(getDomFromRef) => (
>         <>
>             <Input ref={(ref) => getDomFromRef(ref.dom)} />
>         </>
>     )}
> </Trigger>;
> ```
> 
> 另外可能还需要在文档里注明，都有哪些组件 ref 返回的不是个 dom，比如 Input

我们本周试一下

**Comment by hu-tu-yi-shi** [comment]
https://github.com/arco-design/arco-design/search?q=findDOMNode&type=code
 挺多的

**Comment by hu-tu-yi-shi** [comment]
这个bug快一年了，有没有哪个大哥处理一下

**Comment by hu-tu-yi-shi** [comment]
关于支持react18， 你们可以参考 https://github.com/ant-design/ant-design/issues/26136

**Comment by yinkaihui** [comment]
findDOMNode 尝试过一个版本，考虑到业务使用方式， 目前还没有办法完全平滑兼容

**Comment by pollux2021** [comment]
这个东西还没进展吗？ 过去蛮久了

**Comment by ekil1100** [comment]
有什么进展吗？碰到的困难是什么？ @PengJiyuan

**Comment by ranlix** [comment]
> findDOMNode 尝试过一个版本，考虑到业务使用方式， 目前还没有办法完全平滑兼容

请问这个问题已经一年半了，有什么进展吗？😭

**Comment by Mark24Code** [comment]
关注。。。这个 warn 太占位置了

**Comment by Jungzl** [comment]
```typescript
const consoleError = console.error.bind(console)
console.error = (errObj, ...args) =>
  !(isDev && (typeof errObj === 'string' || typeof errObj.message === 'string') && args.includes('findDOMNode')) &&
  consoleError(errObj, ...args)
```
https://github.com/ant-design/ant-design/issues/26136#issuecomment-965111940

**Comment by PengJiyuan** [comment]
Close：https://github.com/arco-design/arco-design/issues/1447

**Comment by danaoairuike** [comment]
这个issue有2年多了


> @PengJiyuan 在源码里搜了一下，findDOMNode 出现的次数也不算太多，看起来都能处理一下
> 
> 1. 有个叫 getNodeHeight 的函数，看了下传参已经是一个 HTMLElement 了，没必要用 findDOMNode 了
> 2. ResizeObserverComponent 是个内部的组件，应该可以直接换成读 children 的 ref
> 3. 有些组件里用到了 findDOMNode(getPopupContainer())，这里可以判断下，如果 getPopupContainer 返回的是个 ReactElement 再使用 findDOMNode，并在文档里提示
> 4. 最后一个就是 Trigger 了，首先里面有一半是 findDOMNode(this.triggerRef)，可以直接去掉。然后麻烦的就是获取 children 的 dom 了
> 
> 目前 mui 的做法是强制 children 必须有 ref
> 
> 这里感觉可以改成先读取 children 的 ref，如果拿到的不是 HTMLElement 的话，再使用 findDOMNode
> 
> 对于像 Input 这样 ref 返回的不是个 dom 的情况，要么是让用户自己用一个新的组件包一层，要么是 Trigger 支持下自定义从 ref 获取 dom
> 
> 比如这样
> 
> 方式1，缩进会少一点
> 
> ```js
> <Trigger getDomFromRef={ref => ref.dom}>
>     <Input />
> </Trigger>
> ```
> 
> 方式2，缩进有点多，但是更灵活，虽然我也不知道什么情况下会这么写，明明把 Trigger 直接写里面就好
> 
> ```js
> <Trigger>
>     {(getDomFromRef) => (
>         <>
>             <Input ref={(ref) => getDomFromRef(ref.dom)} />
>         </>
>     )}
> </Trigger>;
> ```
> 
> 另外可能还需要在文档里注明，都有哪些组件 ref 返回的不是个 dom，比如 Input

这个可以解决问题，为何不采纳？

**Comment by guiyumin** [comment]
> https://github.com/arco-design/arco-design/search?q=findDOMNode&type=code 挺多的

7个file而已， 都2024年了

**Comment by lyspro** [comment]
this issue has been addressed for the originally reported built-in Menu + StrictMode scenario in recent versions of arco-design
If you are still seeing warnings, please:
- Ensure you're on the latest version of `@arco-design/web-react`
- Provide a minimal reproduction with your exact version — we'll investigate further.
